File: Queue/Priority Queue/요격 시스템.cpp
// 프로그래머스 Lv.2 요격 시스템

#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

priority_queue<int, vector<int>, greater<int>> pq;

int solution(vector<vector<int>> targets) {
    int answer = 1;
    sort(targets.begin(), targets.end());
    
    for(int i=0; i<targets.size(); i++){
        if(pq.empty())
            pq.push(targets[i][1]);
        
        else if(pq.top() <= targets[i][0]){ // 요격 미사일이 추가로 필요하다면
            answer++;
            while(!pq.empty()) // 지금까지 요격한 폭격 미사일 목록 삭제
                pq.pop();
            
            pq.push(targets[i][1]); // 앞으로 요격해야 할 폭격 미사일 추가
        }
        else
            pq.push(targets[i][1]);
    }
    return answer;
}

File: Queue/Priority Queue/요격 시스템.java
// 프로그래머스 Lv.2

import java.util.*;

class Position {
    int x, y;
    
    Position(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

class PosComparator implements Comparator<Position> {
    
    @Override
    public int compare(Position p1, Position p2) {
        if(p1.x == p2.x) {
            return p1.y - p2.y;
        } else {
            return p1.x - p2.x;
        }
    }
}

class Solution {
    public int solution(int[][] targets) {
        int answer = 0, threshold = 0;
        PriorityQueue<Position> targetList = new PriorityQueue<>(1, new PosComparator());
        
        for(int i=0; i<targets.length; i++) {
            int x = targets[i][0];
            int y = targets[i][1];
            
            targetList.offer(new Position(x, y));
        }
        
        while(!targetList.isEmpty()) {
            Position p = targetList.poll();
            
            if(p.x >= threshold) {
                threshold = p.y;
                answer++;
            }
            
            if(p.y < threshold) {
                threshold = p.y;
            }
        }
        return answer;
    }
}
